test/illuminate/Illuminate.sol

```diff
@@ -2,48 +2,95 @@
 
-import "../tokens/ZcToken.sol";
+import "./Interfaces.sol";
+import "./ZcToken.sol";
+import "./Safe.sol";
 
 contract Illuminate {
+  /// @notice for any given illuminate deployment the available principles are contained here
+  enum Principals {
+    Illuminate,
+    Swivel,
+    Yield,
+    Element,
+    Pendle,
+    Tempus,
+    Sense,
+    Apwine
+  }
+
+  /// markets are defined by a market pair which point to a fixed length array of principal token addresses.
+  /// the principal tokens those addresses represent correspond to their Principals enum value, thus the
+  /// array should be ordered in that way
+  mapping (address => mapping (uint256 => address[8])) public markets;
+
   address public admin;
-  // a marketplace contracts has many markets
-  // a market is a specific set of Principals (token addresses)
-  // a market is identified by a market-pair which is underlying:maturity (address:uint256)
-  address public marketPlace;
-  // Illuminite essenttially dictates the interface, Iilluminate, which is
-  // * mint
-  // * lend
-  // * redeem
-  // * ...
-  // in order to adapt Principals to this interface the lender keeps interfaces for any/all lending
-  address public lender;
-  // and the same goes for redeeming
+  /// @notice address of the deployed lender contract
+  address public lender; // TODO add authorized methods to set these
+  /// @notice address of the deployed redeemer contract
   address public redeemer;
 
-  event Mint(uint8 principal, address indexed underlying, uint256 indexed maturity, uint256 amount);
+  event CreateMarket(address indexed underlying, uint256 indexed maturity);
+
+  // TODO move to lender.sol
+  // event Mint(uint8 principal, address indexed underlying, uint256 indexed maturity, uint256 amount);
 
-  /// @param m address of a deployed market contract to use
-  /// @param r address of a deployed relayer contract to use
-  constructor(address m, address r) {
+  constructor() {
     admin = msg.sender;
   }
 
+  /// @param u underlying
+  /// @param m maturity
+  /// @param t principal token addresses for this market minus the illuminate principal (which is added here)
+  /// @param n name for the illuminate token
+  /// @param s symbol for the illuminate token
+  /// @param d decimals for the illuminate token
+  function createMarket(address u, uint256 m, address[7] memory t, string calldata n, string calldata s, uint8 d) external authorized(admin) returns (bool) {
+    require(markets[u][m][uint256(Principals.Illuminate)] == address(0), 'market exists'); 
+
+    // deploy an illuminate token with this new market NOTE: ATM is using name as symbol args
+    address iToken = address(new ZcToken(u, m, n, s, d));
+
+    // the market will have the illuminate principal as its zeroth item, thus t should have Principals[1] as [0]
+    // TODO we could choose to put illuminate last in
+    address[8] memory market = [iToken, t[0], t[1], t[2], t[3], t[4], t[5], t[6]];
+
+    uint256 max = 2**256 - 1;
```

Comment:
We should store this value as a constant in `Constants.sol`. 
e.g. `uint256 MAX_UINT256 = uint256(-1)`
This will be slightly more efficient than computing the maximum value via exponentiation.

Reply:
making a constants file not a pattern we've used before -- typically we'd just make an actual constant in the contract if it's used a bunch. don't think this is? i can take a look tho